src/transformers/models/pop2piano/tokenization_pop2piano.py

```diff
@@ -0,0 +1,794 @@
+            Path to the tokenizer file which contains vocabulary for "NOTE" tokens.
+        vocab_velocity_file (`str`):
+            Path to the tokenizer file which contains vocabulary for "VELOCITY" notes.
+        vocab_special_file (`str`):
+            Path to the tokenizer file which contains vocabulary for "SPECIAL" tokens.
+    """
+
+    model_input_names = ["token_ids"]
+    vocab_files_names = VOCAB_FILES_NAMES
+    pretrained_vocab_files_map = PRETRAINED_VOCAB_FILES_MAP
+
+    def __init__(
+        self,
+        vocab_params_file,
+        vocab_time_file,
+        vocab_note_file,
+        vocab_velocity_file,
+        vocab_special_file,
+        unk_token="-1",
+        eos_token="1",
+        pad_token="0",
+        bos_token="2",
+        **kwargs,
+    ):
+        unk_token = AddedToken(unk_token, lstrip=False, rstrip=False) if isinstance(unk_token, str) else unk_token
+        eos_token = AddedToken(eos_token, lstrip=False, rstrip=False) if isinstance(eos_token, str) else eos_token
+        pad_token = AddedToken(pad_token, lstrip=False, rstrip=False) if isinstance(pad_token, str) else pad_token
+        bos_token = AddedToken(bos_token, lstrip=False, rstrip=False) if isinstance(bos_token, str) else bos_token
+
+        super().__init__(
+            unk_token=unk_token,
+            eos_token=eos_token,
+            pad_token=pad_token,
+            bos_token=bos_token,
+            **kwargs,
+        )
+
+        self.encoder = {}
+
+        # Load all vocab files
+        with open(vocab_params_file, "rb") as file:
+            self.encoder["TOKEN_PARAMS_VOCAB"] = json.load(file)
+        with open(vocab_time_file, "rb") as file:
+            self.encoder["TOKEN_TIME_VOCAB"] = json.load(file)
+        with open(vocab_note_file, "rb") as file:
+            self.encoder["TOKEN_NOTE_VOCAB"] = json.load(file)
+        with open(vocab_velocity_file, "rb") as file:
+            self.encoder["TOKEN_VELOCITY_VOCAB"] = json.load(file)
+        with open(vocab_special_file, "rb") as file:
+            self.encoder["TOKEN_SPECIAL_VOCAB"] = json.load(file)
+
+        # create mappings for token_ids to note or the reverse of self.encoder
+        self.decoder = {}
+        self.decoder["TOKEN_TIME_VOCAB"] = {v: k for k, v in self.encoder["TOKEN_TIME_VOCAB"].items()}
+        self.decoder["TOKEN_NOTE_VOCAB"] = {v: k for k, v in self.encoder["TOKEN_NOTE_VOCAB"].items()}
+        self.decoder["TOKEN_VELOCITY_VOCAB"] = {v: k for k, v in self.encoder["TOKEN_VELOCITY_VOCAB"].items()}
+        self.decoder["TOKEN_SPECIAL_VOCAB"] = {v: k for k, v in self.encoder["TOKEN_SPECIAL_VOCAB"].items()}
+
+    @property
+    def vocab_size(self):
+        """Returns the vocabulary size of the tokenizer."""
+        return (
+            len(self.encoder["TOKEN_TIME_VOCAB"])
+            + len(self.encoder["TOKEN_NOTE_VOCAB"])
+            + len(self.encoder["TOKEN_VELOCITY_VOCAB"])
+            + len(self.encoder["TOKEN_SPECIAL_VOCAB"])
+        )
+
```

Comment:
I didn't add this since it conflicts with `self.vocab_size_special` also if we are implementing this for `token special` we might need to do for all tokens. Should I add this method for all tokens then? What do you think?

Reply:
Yes, for all the 4 vocabularies 😉

Reply:
(I mean let's remove the self.xxxx and add the properties

Reply:
Hello @ArthurZucker , I am sorry there is one problem, I forgot to mention - 

Actually `self.vocab_size_time`, `self.vocab_size_note`, `self.vocab_size_velocity` and `self.vocab_size_special`  are different from the `self.token_special_encoder`, `self.token_note_encoder`, `self.token_time_encoder` and  `self.token_velocity_encoder` (the dicts that we use to retrieve the token-id values).

The `self.vocab_size_time` does NOT indicate the size of `self.token_time_encoder` vocab. And this logic goes for other token types too.

The `self.vocab_size_time` and others were used in the official implementation to get the `tokenize` and `detokenize` methods([here](https://github.com/sweetcocoa/pop2piano/blob/fac11e8dcfc73487513f4588e8d0c22a22f2fdc5/midi_tokenizer.py#L34)). Since the official tokenizer didn't have any vocab I manually created them(`XX_encoder`) by getting the all possible values of model output(as described [here](https://github.com/huggingface/transformers/pull/21785#discussion_r1283152996)). In fact all vocabs have the same shape as 2400(since its the no of possible outputs that the model produces).

So we must get the original values of `self.vocab_size_time`, `self.vocab_size_note`, `self.vocab_size_velocity` and `self.vocab_size_special` from the `__init__` instead of calculating using `len(self.token_XX_encoder)`.

Do you think that it's better to change the design to the previous one where there were no vocabs(only `self.vocab_size_XX`) to remove this confusion? Or go with this one(in this case all the vocab sizes will be from the `__init__`)?

Reply:
Hi @ArthurZucker, yes this makes sense to me! 

So, how about we also add the token information and make a single vocab like - 
```
'TOKEN_NOTE_126' : 130,
'TOKEN_NOTE_127' : 131,
'TOKEN_SPECIAL_0' : 0,
'TOKEN_SPECIAL_1' : 1,
'TOKEN_SPECIAL_2' : 2,
'TOKEN_SPECIAL_3' : 3,
'TOKEN_TIME_0' : 134,
'TOKEN_TIME_1' : 135,
```
("`<token type>`_`<token value>`"  :  `token_ids`)

In this way we can remove the indexing needed for determining the token types(for the method `_convert_id_to_token`). Now there will be 1) a single vocab like other tokenizers and 2) almost no indexing.  

What do you think?

Reply:
Yes that's a lot simpler and should help understanding from just reading the vocab file! 😉

Reply:
Thanks for the clarification!

Reply:
Okk, then I will save the decoder as `vocab.json` and I wanted to ask one more question please - 

what do you think of the vocab size methods? I mean that, in this case we don't need to implement the `vocab_size_special`, `vocab_size_note`, `vocab_size_time` and `vocab_size_velocity` since we are not using those values anywhere in the tokenizer file (only `vocab_size_note` is used once in the method `relative_tokens_ids_to_notes`).

I believe we only need the `vocab_size` method, right?

Reply:
yes we only need the vocab size

Reply:
+1 on this tokenizer design - think it's a very nice this way of organising the tokenizer, and much cleaner than having separate vocabs / methods for each of the token types